prio/encrypt.c

```diff
@@ -0,0 +1,254 @@
+
+// Use Curve25519
+#define CURVE_OID_TAG SEC_OID_CURVE25519
+#define CURVE25519_KEY_LEN 32
+
+// Use 96-bit IV
+#define GCM_IV_LEN_BYTES 12
+// Use 128-bit auth tag
+#define GCM_TAG_LEN_BYTES 16
+
+#define PRIO_TAG "PrioPacket"
+#define AAD_LEN (strlen (PRIO_TAG) + CURVE25519_KEY_LEN + GCM_IV_LEN_BYTES)
+
+static SECStatus 
+derive_dh_secret (PK11SymKey **shared_secret, PrivateKey priv, PublicKey pub)
+{
+  if (priv == NULL) return SECFailure;
+  if (pub == NULL) return SECFailure;
+  if (shared_secret == NULL) return SECFailure;
+
+  SECStatus rv = SECSuccess;
+  *shared_secret = NULL;
+
+  P_CHECKA (*shared_secret = PK11_PubDeriveWithKDF (priv, pub, PR_FALSE,
+        NULL, NULL, CKM_ECDH1_DERIVE, CKM_AES_GCM, 
+        CKA_ENCRYPT | CKA_DECRYPT, 16,
+        CKD_SHA256_KDF, NULL, NULL));
+
+
+cleanup:
+  return rv;
+}
+
+SECStatus
+PublicKey_import (PublicKey *pk, const unsigned char *data, unsigned int dataLen)
+{
+  SECStatus rv = SECSuccess;
+  PrivateKey eph_priv = NULL;
+  *pk = NULL;
+
+  if (dataLen != CURVE25519_KEY_LEN)
+    return SECFailure;
+
+  // TODO: Horrible hack. We should be able to import a public key
+  // directly without having to generate one from scratch.
+  P_CHECKC (Keypair_new (&eph_priv, pk));
+  memcpy ((*pk)->u.ec.publicValue.data, data, CURVE25519_KEY_LEN);
+
+cleanup:
+  if (rv != SECSuccess)
+    PublicKey_clear (*pk);
+  PrivateKey_clear (eph_priv);
+  return rv;
+}
+
+SECStatus 
+Keypair_new (PrivateKey *pvtkey, PublicKey *pubkey)
+{
+  if (pvtkey == NULL) return SECFailure;
+  if (pubkey == NULL) return SECFailure;
+
+  SECStatus rv = SECSuccess;
+  SECOidData *oid_data = NULL;
+  *pubkey = NULL;
+  *pvtkey = NULL;
+
+  SECKEYECParams ecp;
+  ecp.data = NULL;
+  PK11SlotInfo *slot = NULL;
+
+  P_CHECKA (oid_data = SECOID_FindOIDByTag (CURVE_OID_TAG));
+  const int oid_struct_len = 2 + oid_data->oid.len;
+
+  P_CHECKA (ecp.data = malloc (oid_struct_len));
+  ecp.len = oid_struct_len;
+  
+  ecp.type = siDEROID;
+
+  ecp.data[0] = SEC_ASN1_OBJECT_ID;
+  ecp.data[1] = oid_data->oid.len;
+  memcpy (&ecp.data[2], oid_data->oid.data, oid_data->oid.len);
+
+  P_CHECKA (slot = PK11_GetInternalSlot ());
+  P_CHECKA (*pvtkey = PK11_GenerateKeyPair(slot, CKM_EC_KEY_PAIR_GEN, &ecp, 
+      (SECKEYPublicKey **)pubkey, PR_FALSE, PR_FALSE, NULL));
+
+cleanup:
+  if (ecp.data)
+    free (ecp.data);
+  if (rv != SECSuccess) {
+    PublicKey_clear (*pubkey);
+    PrivateKey_clear (*pvtkey);
+  }
+  return rv;
+}
+
+void 
+PublicKey_clear (PublicKey pubkey)
+{
+  if (pubkey) 
+    SECKEY_DestroyPublicKey(pubkey);
+}
+
+void 
+PrivateKey_clear (PrivateKey pvtkey)
+{
+  if (pvtkey) 
+    SECKEY_DestroyPrivateKey(pvtkey); 
+}
+
+const SECItem *
+PublicKey_toBytes (const_PublicKey pubkey)
+{
+  return &pubkey->u.ec.publicValue;
+}
+
+unsigned int
+PublicKey_encryptSize (unsigned int inputLen)
+{
+  // public key, IV, tag, and input
+  return CURVE25519_KEY_LEN + GCM_IV_LEN_BYTES + GCM_TAG_LEN_BYTES + inputLen;
```

Comment:
This can overflow and thus cause badness. You should generally check lengths before adding them up to make sure they not getting too large.

Reply:
Oh yes, thanks for pointing this out. I will change `PublicKey_encryptSize` to return an error flag to catch values so large as to cause integer overflow.